Library/Homebrew/cmd/tap.rb

```diff
@@ -17,8 +17,10 @@ def tap
         tap.install :clone_target => ARGV.named[1],
                     :full_clone   => ARGV.include?("--full"),
                     :quiet        => ARGV.quieter?
-      rescue TapAlreadyTappedError => e
-        opoo e
+      rescue TapRemoteMismatchError => e
+        odie e
+      rescue TapAlreadyTappedError, TapAlreadyUnshallowError => e
+        $stderr.puts e unless ARGV.quieter?
```

Comment:
Yeah, I think regular output and not a warning makes sense for idempotent operations. If you're using things right, you should be able to go through a process without getting warnings. Calling `brew tap --full foo` means to make sure that the `foo` tap is installed and is not shallow. If that's already the case, then the "already tapped/unshallow" output is just confirming things are in the state the user requested. There's no oddness they need to investigate, and there's no different actions they should have done, and there's nothing about the state change from shallow to unshallow that the user probably cares about: they just want it unshallow, regardless of its initial state.

I don't think there even is a way to avoid the warning using just `brew` commands, as I don't see a `brew tap` variant that will output whether or not a tap is unshallow. Seems like warnings should be for things that indicate potential problems and require user attention. This case doesn't. It's like an `rsync` operation: if all the files are already at the destination, it doesn't copy anything, but it doesn't warn about anything.